### PR TITLE
fix: resolve etxtbsy issue(file busy error) at spawn

### PR DESCRIPTION
**Pull request by shivamku-BS**
## Proposed changes

Fixes a binary file busy (ETXTBSY/EBUSY) error that occurred when the BrowserStack CLI binary was being executed while a new version was being written to the same file path during spawn.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Backport Request

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 1]
```diff
@@ -372,6 +385,26 @@ export class CLIUtils {
...
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        } catch (err: any) {
+            if (BINARY_BUSY_ERROR_CODES.includes(err.code)) {
+                logger.debug(`Binary is busy: ${binaryPath}`)
+                return true
+            }
+            logger.debug(`Error checking if binary is busy: ${err.message}`)
+            return false
+        }
+    }
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:413]: The 'r+' probe also surfaces Windows sharing violations as EBUSY/EPERM, which are in BINARY_BUSY_ERROR_CODES. The spawn-retry layer is the real safety net across platforms; this probe is best-effort.

File: packages/wdio-browserstack-service/src/cli/index.ts [hunk 2]
```diff
@@ -206,52 +206,105 @@ export class BrowserstackCLI {
...
+                settle(reject, err)
             })
 
-            this.process!.on('close', (code: number) => {
+            this.process.on('close', (code: number) => {
                 if (code !== 0) {
-                    reject(new Error(`CLI process exited with code ${code}`))
+                    settle(reject, new Error(`CLI process exited with code ${code}`))
                 }
             })
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/index.ts:324]: Have resolved that

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 3]
```diff
@@ -703,37 +736,91 @@ export class CLIUtils {
...
+
+                    writeStream.on('error', (writeErr) => {
+                        fsp.unlink(tempPath).catch(() => {})
+                        zipfile.close()
+                        reject(writeErr as Error)
+                    })
 
-                        if (entry.fileName === binaryName) {
+                    // 'finish' fires on successful flush only; 'error' path won't reach here.
+                    writeStream.on('finish', async () => {
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts]: resolved that

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 4]
```diff
@@ -703,37 +736,91 @@ export class CLIUtils {
...
+                    const finalPath = path.join(cliDir, entry.fileName)
+                    const tempPath = path.join(cliDir, `${entry.fileName}.tmp.${process.pid}`)
+
+                    const writeStream = fs.createWriteStream(tempPath)
+
+                    writeStream.on('error', (writeErr) => {
+                        fsp.unlink(tempPath).catch(() => {})
+                        zipfile.close()
+                        reject(writeErr as Error)
+                    })
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:804]: have resolved

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 5]
```diff
@@ -703,37 +736,91 @@ export class CLIUtils {
...
 
-                        if (entry.fileName === binaryName) {
+                    // 'close' fires after the fd is closed; safe for fsp.rename on Windows (where 'finish' may fire before fd release).
+                    writeStream.on('close', async () => {
+                        try {
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:810]: have resolved

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 6]
```diff
@@ -710,37 +743,91 @@ export class CLIUtils {
...
+                            zipfile.close()
                             reject(zipErr as Error)
                         }
+                        return
+                    }
+
+                    // Binary entry: extract to PID-scoped temp file, chmod, atomic rename inline.
+                    // Prevents ETXTBSY/EBUSY: the file being executed is never the file being written.
+                    const finalPath = path.join(cliDir, entry.fileName)
+                    const tempPath = path.join(cliDir, `${entry.fileName}.tmp.${process.pid}`)
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:794]: have resolved

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 7]
```diff
@@ -710,37 +743,91 @@ export class CLIUtils {
...
+                                directStream.end()
+                                directStream.on('close', () => zipfile.readEntry())
                             })
-                            readStream.pipe(writeStream)
+                            readStream.pipe(directStream)
                         } catch (zipErr) {
+                            zipfile.close()
                             reject(zipErr as Error)
                         }
+                        return
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:788]: have resolved

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 8]
```diff
@@ -710,37 +743,108 @@ export class CLIUtils {
...
+                        const readStream = await openReadStreamPromise(entry)
+                        readStream.on('end', function () {
+                            writeStream.end()
+                        })
+                        readStream.pipe(writeStream)
+                    } catch (zipErr) {
+                        fsp.unlink(tempPath).catch(() => {})
+                        zipfile.close()
+                        reject(zipErr as Error)
                     }
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:848]: as per the scope for this PR. This promise only rejects if the zip header is completely malformed. If that happens, the test run is doomed anyway, so we're fine relying on the OS process exit to clean up the dangling temp file descriptor in this extreme edge case.

File: packages/wdio-browserstack-service/src/cli/cliUtils.ts [hunk 9]
```diff
@@ -710,37 +743,108 @@ export class CLIUtils {
...
+
+                    const openReadStreamPromise = promisify(
+                        zipfile.openReadStream,
+                    ).bind(zipfile)
+                    try {
+                        const readStream = await openReadStreamPromise(entry)
+                        readStream.on('end', function () {
+                            writeStream.end()
+                        })
+                        readStream.pipe(writeStream)
```
[the remark this answers is not in the document]
Comment (shivamku-BS) [reply packages/wdio-browserstack-service/src/cli/cliUtils.ts:843]: Not required. We've eliminated concurrent write corruption by scoping the zip download to the process ID. Any remaining mid-stream read errors would strictly be from network drops or disk issues, which are out of scope for this specific race-condition PR. also if such situation  occurs we have cleanup logic for orphan binaries.

**Review by kamal-kaur04 — APPROVED** [review]
